server/caas/api.js

```diff
@@ -0,0 +1,324 @@
+
+  var helper = require('sendgrid').mail;
+  var crypto = require('crypto');
+
+
+  app.use(bodyParser.json()); // for parsing application/json
+
+  var pool = mysql.createPool({
+    host     : 'localhost',
+    user : 'root',
+    password : 'asdf',
+    database : 'ui',
+  });
+
+  app.use('/register-new', function(req, res, next) {
+    generateAndInsertToken(null, req.body.name, req.body.email, "create", function(error, results, fields, token) {
+      if (error) {
+        res.status(500).send({error: error});
+      } else {
+        sendVerificationEmail(req.body.email, req.body.name, req.headers.origin, token, function(error, resp) {
+          if (error) {
+            res.status(500).send({error: error});
+          } else {
+            res.status(200).json({success: 'Email sent'});
+          }
+        });
+      }
+    });
+  });
+
+
+  app.use('/verify-token', function(req, res, next) {
+    pool.query('SELECT * FROM challenge WHERE token = ? AND created > DATE_SUB(NOW(), INTERVAL 24 HOUR)', [req.body.token], function(error, results, fields) {
+      if (error) {
+        res.status(500).send();
+      } else {
+        if (results.length) {
+          var row = results[0];
+          var out = {
+            email: row.email,
+            name: row.name,
+          };
+          res.status(200).send(out);
+        } else {
+          res.status(400).send({error: 'Token not found.'});
+        }
+      }
+    });
+  });
+
+  app.use('/create-user', function(req, res, next) {
+    var user = req.body;
+    // {kind: "user", type: "account", name: "wes"}
+    var account = {
+      type: 'user',
+      name: user.name
+    };
+    // {type: "password", publicValue: "wes1@rancher.com", secretValue: "asdf", accountId: "1a28"}
+    var credential = {
+      type: 'password',
+      publicValue: user.email,
+      secretValue: user.pw,
+      accountId: null
+    };
+    var url = config.apiServer + config.apiEndpoint;
+
+    pool.query('SELECT * FROM challenge WHERE token = ? AND created > DATE_SUB(NOW(), INTERVAL 24 HOUR)', [user.token], function(error, results, fields) {
+      if (error) {
+        res.status(500).send();
+      } else {
+        if (results.length) {
+          var accountModel = null;
+          var credentialModel = null;
+          newRequest({
+            url: url + '/account',
+            method: 'POST',
+            body: account
+          }, function(response, body) {
+
+            accountModel = body;
+            credential.accountId = body.id;
+
+            newRequest({
+              url: url + '/passwords',
+              method: 'POST',
+              body: credential
+            }, function(response, body) {
+              credentialModel = body;
+              removeUserFromTable(user.email, function(results, fields) {
+                getTokenForLogin(user.email, user.pw, function(response, body) {
+                  res.cookie('token', body.jwt, {secure: true});
```

Comment:
[req.secure](https://expressjs.com/en/4x/api.html#req.secure) is a property that tells you if the request came in via SSL.  You only want the `secure` property on the cookie fi the request is also secure.